Integrate eyecite 2.1 (part 1)

mattdahl opened the pull request:
This PR is nearly complete but I will add more text here when it is ready to merge.

mlissner commented on cl/citations/annotate_citations.py:
```diff
@@ -0,0 +1,100 @@
+        if not isinstance(citation, NonopinionCitation)
+    ]
+
+    # STEP 1: Generate the string annotations to add
```
Feels like pulling these two steps into their own functions might help simplify a bit.

mlissner commented on .github/workflows/lint.yml:
```diff
@@ -62,7 +62,8 @@ jobs:
           -p cl.simple_pages \
-          -p cl.recap_rss
+          -p cl.recap_rss \
+          -p cl.citations
```
Nice, this is great, but could you alphabetize it?

mattdahl replied:
Ah yes.

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -42,11 +42,15 @@ def make_name_param(defendant, plaintiff=None):
 
-def reverse_match(conn, results, citing_doc):
+def reverse_match(
+    conn: ExtraSolrInterface, results: List, citing_doc: Opinion
```
Might be good to say what this is a `List` of, but in CourtListener I'm only looking for progress not perfection wrt types.

mattdahl replied:
Yes, I'll specify that.

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -83,26 +92,23 @@ def case_name_query(conn, params, citation, citing_doc):
 
 
-def get_years_from_reporter(citation):
+def get_years_from_reporter(citation: CitationBase) -> Tuple[int, int]:
```
FWIW, I just learned you can just do `-> (int, int)`, instead of the `Tuple` business.

mattdahl replied:
Is that preferable?

mlissner replied:
Actually, I was wrong. Nevermind.

mlissner commented on cl/citations/tasks.py:
```diff
@@ -240,6 +145,11 @@ def find_citations_for_opinion_by_pks(
                 ]
             )
 
+            # Delete the memoized opinion data (not saved in DB)
+            del opinion.source_text
+            del opinion.cleaned_text
+            del opinion.source_is_html
+
```
I wouldn't put this in the transaction block, but in any case, seeing this does make me a bit uncomfortable with the design. Is it so much better to pass the opinion object around instead of something else that we could throw away after?

This cleanup is here to fix something a different function did, so it's pretty messy.

mattdahl replied:
What if I just removed these lines? I don't think they're actually necessary to have the model save correctly. These memoized fields won't be propagated to the database in any scenario.

mlissner replied:
Yeah, the only risk is if we later create fields with those names. Pretty unlikely.

mlissner commented on cl/citations/tests.py:
```diff
@@ -212,46 +210,42 @@ def test_make_html_from_matched_citation_objects(self) -> None:
+                opinion=opinion,
+                citations=citations,
+            )
             self.assertEqual(
```
It'd be nice to use self.subTest for any of these with lists of test cases. Makes them a lot easier to deal with.

mattdahl replied:
Interesting trick, I will do that.

mlissner replied:
Got it from Jack in eyecite. It's new in py3.

mlissner commented on pyproject.toml:
```diff
@@ -91,7 +91,7 @@ scorched = {git = "https://github.com/freelawproject/scorched.git"}
 ipython = "^7.21.0"
 gunicorn = "^20.0.4"
-eyecite = "<2.0.0"
+eyecite = {git = "https://github.com/freelawproject/eyecite.git", rev = "0595c93"}
```
Nice. Let's cut a release for this once your counting PR is merged?

mattdahl replied:
Actually, I think it would be easier/safer to do the counting refactor as another PR after this one. I'm also not sure how long it will take me to do the tests for the counting PR in eyecite. But we can also hold off here until that's finalized, it doesn't matter much to me.

mlissner replied:
So long as somebody is thinking about this, I'm happy.

mlissner replied:
OK, I think this is the last piece that's missing before we merge this, right? Otherwise, we're good to go here as far as I can tell.

mattdahl replied:
Yeah, what did we decide here? Do you want to deploy a versioned tag of eyecite versus one just referencing a commit?

mlissner reviewed (changes requested):
This looks great to me. I haven't checked each test manually, but it looks like you gave them each some care. I made a few comments, but it seems like as usual you've got all the right ideas, maybe save one. :)

Nice work getting mypy done too. It's a bit of work, but it sure makes reviews easier and makes skimming the code later much easier.

mattdahl commented on cl/citations/annotate_citations.py:
```diff
@@ -0,0 +1,107 @@
+
+def generate_annotations(
+    citations: List[CitationBase],
+) -> List[List]:
```
I want the return signature here to be `List[List[Union[Tuple[int, int], str]]]`, i.e., a list of lists where the inner list can contain either strings or tuples of ints. However, this does not seem to work, and I'm not sure what I'm doing wrong.

mlissner replied:
Eh, let's not force it. If the above passes, I'm sure we can dial it in further later when we're better at mypy.

mattdahl commented on cl/citations/fixtures/opinions_matching_citations.json:
```diff
@@ -379,7 +379,7 @@
             "extracted_by_ocr": false,
             "author": 2,
-            "plain_text": "Blah blah Foo v. Bar 1 U.S. 1, 77 blah blah. Asdf asdf Qwerty v. Uiop 2 F.3d 2, 555. Also check out Foo, 1 U.S. 1, at 99. Then let's cite Qwerty, supra, at 666. See also Foo, supra, at 101 as well. Another full citation is Lorem v. Ipsum 1 U. S. 50. Quoting Qwerty, “something something”, 2 F.3d 2, at 559. This case is similar to Fake, supra, and Qwerty supra, as well. T ...
+            "plain_text": "Blah blah Foo v. Bar 1 U.S. 1, 77 blah blah. Asdf asdf Qwerty v. Uiop 2 F.3d 2, 555. Also check out Foo, 1 U.S. at 99. Then let's cite Qwerty, supra, at 666. See also Foo, supra, at 101 as well. Another full citation is Lorem v. Ipsum 1 U. S. 50. Quoting Qwerty, “something something”, 2 F.3d 2, at 559. This case is similar to Fake, supra, and Qwerty supra, as well. This ...
```
This change is not necessary to make the tests pass; it was just a correction to the existing test case because I had previously said that there were a certain number of full case citations and a certain number of short case citations in this test string. Because both get counted, the test does not not hinge on which is which, but I just wanted to be consistent with the documentation.

mattdahl commented on cl/citations/match_citations.py:
```diff
@@ -83,26 +94,23 @@ def case_name_query(conn, params, citation, citing_doc):
-        if hasattr(reporter_dates["end"], "year"):
-            end_year = reporter_dates["end"].year
+
+    edition_guess = citation.edition_guess
```
I'm not entirely sure this is correct, so I want to flag this as requiring a closer look. The previous logic with the four-levels of nesting in the `REPORTERS` variable confused me, but I think we can just grab the same information now using the `edition_guess` attribute?

mlissner replied:
This looks right to me.

mattdahl commented on cl/citations/match_citations.py:
```diff
@@ -120,7 +128,7 @@ def match_citation(citation, citing_doc=None):
     if citing_doc is not None:
         # Eliminate self-cites.
-        main_params["fq"].append("-id:%s" % citing_doc.pk)
+        cast(List[str], main_params["fq"]).append("-id:%s" % citing_doc.pk)
```
This `cast()` call is a bit ugly but I found it necessary because mypy doesn't know whether this particular parameter (`fq`) maps to a str, int, or a list of strings.

mattdahl replied:
(ditto for below)

mlissner replied:
`SearchParam` will fix this too.

mattdahl commented on cl/citations/match_citations.py:
```diff
@@ -218,7 +232,7 @@ def get_citation_matches(
                     if (
                         citation.reporter == c.reporter
-                        and citation.volume == c.volume
+                        and citation.volume == str(c.volume)
```
In eyecite volumes are now strings rather than ints (see https://github.com/freelawproject/eyecite/pull/12), so there is some mismatch between eyecite's citation abstraction and CL's. One solution is to migrate all the existing citations in CL from int to str, but that seems dangerous to me and would at least require a separate PR. In the meantime, I simply decided to coerce the volume to str here to do a proper comparison.

mattdahl commented on cl/citations/tasks.py:
```diff
@@ -205,10 +110,13 @@ def find_citations_for_opinion_by_pks(
                 opinion_ids_to_update.add(matched_opinion.pk)
 
+        # Generate the citing opinion's new HTML with inline citation links
+        opinion.html_with_citations = create_cited_html(opinion, citations)
```
It seemed to me that this shouldn't actually be in the transaction block at all.

mattdahl commented on cl/citations/tests.py:
```diff
@@ -37,106 +37,97 @@ def test_make_html_from_plain_text(self) -> None:
+             'inline"> asdf</pre>'),
 
             # Using a variant format for U.S. (Issue #409)
-            ('asdf 22 U. S. 33 asdf', full_citation_html),
```
I actually think this was a bug before -- isn't our principle to not change the underlying text of the opinion at all? As this test suggests, we seemed to have been using the erstwhile canonical `reporter` attribute instead of the original `reporter_found` when constructing the citation HTML.

mlissner replied:
Yeah, better not to mess with it, though I could see my past self thinking we could make it better by using the typo-free version of the reporter.

mattdahl commented on cl/citations/tests.py:
```diff
@@ -37,106 +37,97 @@ def test_make_html_from_plain_text(self) -> None:
             # Full citation across line break
             ('asdf John v. Doe, 123\nU.S. 456, upholding foo bar',
              '<pre class="inline">asdf John v. Doe, </pre><span class="'
-             'citation no-link"><span class="volume">123</span>\n<span class='
```
As you will see, the HTML for all of these is simplified (no more internal spans for volumes, pages, reporters, etc. as I saw no point to them), but the substance of test should hopefully be unchanged.

mlissner replied:
Yep, seems good to me, thanks.

mattdahl commented on cl/citations/tests.py:
```diff
@@ -282,36 +280,37 @@ def test_citation_resolution(self) -> None:
-                             canonical_reporter='U.S.', lookup_index=0,
-                             court='scotus', reporter_index=1,
-                             reporter_found='U.S.')
+                case_citation(volume='1', reporter='U.S.', page='1',
```
As mentioned elsewhere, volumes are now strings instead of ints in eyecite.

mattdahl commented on cl/search/fixtures/test_objects_search.json:
```diff
@@ -99,7 +99,7 @@
       "blocked": false,
       "syllabus": "some rando syllabus",
-      "date_filed": "1795-06-09",
+      "date_filed": "1895-06-09",
```
This was a strange fix, and I did this a few weeks ago, so I unfortunately don't remember what the problem was exactly. At bottom, though, having this fixture being filed at 1795 was not matching properly, so I changed it to 1895. I believe my guess was that this had something to do with recent changes to `reporters_db` -- something about the reporter associated with the citation associated with this opinion not being active in 1795? I don't think I fully figured it out, but this change was necessary to make the tests pass.

mlissner replied:
Thanks for calling this out. This change broke a separate test, but knowing you changed that made fixing that test easy.

mattdahl commented:
@mlissner I believe I addressed all of your previous comments in this new version. I also just added some comments of my own to explain and/or flag parts of the PR that may seem questionable. I think that this is ready for review again, though I would feel more comfortable on my end if I could see that the Docker build and tests were working. (I've tested this locally on my own machine with a Docker image I built myself, but I'm always nervous that things work differently elsewhere.)

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -111,7 +119,7 @@ def match_citation(citation, citing_doc=None):
     # TODO: Create shared solr connection for all queries
     si = ExtraSolrInterface(settings.SOLR_OPINION_URL, mode="r")
-    main_params = {
+    main_params: Dict[str, Union[List[str], str, int]] = {
```
See `SearchParam`

mlissner commented on cl/citations/match_citations.py:
```diff
@@ -129,15 +137,19 @@ def match_citation(citation, citing_doc=None):
             end_year = min(end_year, citing_doc.cluster.date_filed.year)
 
-    main_params["fq"].append(
+    cast(List[str], main_params["fq"]).append(
```
OK, I'm going to let you find the rest of these. :)

mlissner reviewed (changes requested):
So this looks basically good to me. I will try to re-run the tests and see if that makes them pass (I think when I use them, it uses my permissions), but I'm not sure. 

Anyway, this code is complex enough that I don't have a strong grasp of it anymore, but there are a lot of tests, and if you feel good, I'm happy. I think I just had one typing thing that I think would make life tidier (see other comments), but otherwise, let's do it.

mattdahl commented:
@mlissner thanks for creating that `SearchParam` type. I've done one more rebase that incorporates that (and removes those `cast()` calls I had).

mattdahl commented:
@mlissner where did we leave this? I think I would still like to land these changes first before doing a separate PR to refactor the `match_citations` stuff to make use of the new `eyecite.resolve_citations()` function. If you agree, I can rebase this branch to prepare it for merging.

mlissner commented:
Sorry Matt. Yes, if you think the right next step is to merge this one, let's do it. Thanks for picking it up again. 

Let me know if we need a version of eyecite released. I'd much prefer to use released versions over commits.

mattdahl commented:
Sure, if you cut a new version at the current eyecite HEAD, I'll just use that in this PR to save us from having to build another Docker image later.

mlissner commented:
> if you cut a new version at the current eyecite HEAD

Done. I had to fix an issue with the readme before pypi would let it push, and I updated CHANGES.md, but 2.1.0 (I think) is ready to go.

mattdahl commented:
Thanks! This should be rebased and ready to go then. The only addition is 3f038ff; I had to tweak this constant because of the way eyecite now counts tokens.

N.B., also, in order to build my new Docker image I had to comment out the `"recap_document": ["exact", "in"],` lines in `cl/recap/filters.py`. There may be a similar problem on master right now.

mlissner commented on cl/citations/annotate_citations.py:
```diff
@@ -0,0 +1,107 @@
+
+    :param opinion: The Opinion whose text should be parsed
+    """
+    if opinion.html_anon_2020:
+        opinion.source_text = opinion.html_anon_2020
+        opinion.cleaned_text = clean_text(
+            opinion.html_anon_2020, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.html_columbia:
+        opinion.source_text = opinion.html_columbia
+        opinion.cleaned_text = clean_text(
+            opinion.html_columbia, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.html_lawbox:
+        opinion.source_text = opinion.html_lawbox
+        opinion.cleaned_text = clean_text(
+            opinion.html_lawbox, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.html:
+        opinion.source_text = opinion.html
+        opinion.cleaned_text = clean_text(
+            opinion.html, ["html", "all_whitespace"]
+        )
+        opinion.source_is_html = True
+    elif opinion.plain_text:
+        opinion.source_text = opinion.plain_text
+        opinion.cleaned_text = clean_text(
+            opinion.plain_text, ["all_whitespace"]
+        )
+        opinion.source_is_html = False
```
This could be simplified with something like:

```
for attr in ["html_anon_2020", "html_columbia", "html_lawbox", "html"]:
    text = getattr(opinion, attr)
    if text:
        opinion.source_text = text
        opinion.cleaned_text = clean_text(text, ["html", "all_whitespace"])
        opinion.source_is_html = True
        break
else:
    # Didn't hit the break; use plain text
    text = getattr(opinion, "plain_text")
    opinion.source_text = text
    opinion.cleaned_text = clean_text(text, ["all_whitespace"])
    opinion.source_is_html = False    
```

mlissner commented:
OK, this looks basically ready to me except for a couple things. Sorry for the delay in replying:

1. Poetry is upgrading everything instead of just the ones we want. Can you take a stab at only updating the things we want in the lock file? Sorry, I forget how, but I'd start by reverting it and the pyproject file, and then attempting to update just the eyecite and reporters_db projects. I think you can do that by doing `poetry add reporters_db==xyz`

2. I made a little refactoring comment to simplify some code.

3. We need to get tests passing. To that end, I just sent you an invite to the FLP org. I think if you accept that, your next push will run the rests with the needed permissions. 

A note for deploy:

 - [ ] Mike needs to push a new docker image after merging

Thank you Matt. It's been really rewarding to watch you and Jack pull this together and lift our citations tools.

mlissner commented:
I fixed and pushed the poetry stuff. Going to look at refactoring and tests next.

mlissner commented:
The refactoring, poetry, and push of the docker image are done, but regretfully the tests need some work. Do you mind taking a look? I attached the tests as HTML from my IDE, if that's helpful (it's a txt so I could upload it).

I was hoping to merge this just now because I have more dependencies that I'll be merging into master soon, which I expect will conflict. I can work that out though if tests are working here. 

[Test Results - test_everything.txt](https://github.com/freelawproject/courtlistener/files/6504978/Test.Results.-.test_everything.txt)

mattdahl commented:
Thanks for dealing with all of that, and sorry for the broken tests. I think I have fixed many of them in 17cd3d2, though it's hard for me to tell because the full test suite doesn't pass on my local machine normally (for a variety of credential-related reasons, mostly), and even though I joined the FLP org it doesn't seem like the GitHub Action is working for me still :(

One thing that definitely still is broken is some of the tax court tests, which is related to the upgrade to reporterd_db 3.0. I added an issue concerning what I think the problem is (https://github.com/freelawproject/reporters-db/issues/57), but I don't know anything about tax court citation format so I'm not sure what regexes need to be added exactly.

mlissner commented:
OK, thus guy is merged, thank you Matt. A couple notes:

1. I fixed some tests just by ripping them out. I'm lucky to be able to make these calls, but some of the failing tests were tests on cleanup scripts that we had run previously. They didn't look easy to fix, so I nuked them.

2. Another test was failing because it was testing the search interface for something during a date range. You tweaked the fixture, so the query failed. I updated the query. Fixed.

3. I fixed the poetry lock file, which was funky after I rebased on a different upgrade to it (fine). 

That's it! I'm going to go deploy this thing. Fingers crossed.

mlissner commented:
Oh, one other thing. I think your next PR will run properly now that you're part of the org. I think somebody else saw that at some point. It's annoying it doesn't take effect immediately.

mlissner commented:
And it's deployed! So far so good!

mattdahl commented:
Thank you for shepherding this across the finish line! (And sorry for sort of dumping it on you, haha.)

mlissner commented:
No problem at all. It felt like the hard part was done and the rest was getting it over the finish line. So far nothing crashing, which is pretty great.

mattdahl commented:
Yes, that's a minor miracle. I'm working on integrating the `resolve_citations()` stuff next...
